plugin/metrics/handler.go

```diff
@@ -26,7 +26,11 @@ func (m *Metrics) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg
 	rw := dnstest.NewRecorder(w)
 	status, err := plugin.NextOrFailure(m.Name(), m.Next, ctx, rw, r)
 
-	vars.Report(WithServer(ctx), state, zone, rcode.ToString(rw.Rcode), rw.Len, rw.Start)
+	rc := rw.Rcode
+	if !plugin.ClientWrite(status) {
```

Comment:
so *status* is _NOT_ the rcode. Let me re-read your cover letter for the exact problem.
The status is _only_ used to determine if there still is a need to respond to the client (with an error)

Reply:
We need some comments here, as we munging the status into an rcode. It might make sense to 
a) default to SERVFAIL in these cases
b) use "-" as the rcode string as we don't actually know what happened

WDYT?

Reply:
From what I understood from the CoreDNS codebase, if no plugin writes a response, the status (return code from the chain of plugins) is used to construct a DNS message with error https://github.com/coredns/coredns/blob/master/core/dnsserver/server.go#L318, but I agree that without this context it seems like mixing two things in here.

I think out of the two options you listed I prefer the a) as it seems to be the response the DNS client will receive from CoreDNS instance and I would expect that metrics like `coredns_dns_responses_total{rcode}` will be showing proper rcodes the client receives.

so something like would be fine with you?

```
rc := rw.Rcode
// no response was written, fallback to SERVFAL rcode for metrics
if !plugin.ClientWrite(status) {
		rc = dns.RcodeServerFailure
}
```

or should I just add clarifying comment to the current solution?

Reply:
what you have now looks good